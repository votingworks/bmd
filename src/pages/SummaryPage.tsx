import React, { useContext, useState } from 'react'
import { Link, RouteComponentProps } from 'react-router-dom'
import styled from 'styled-components'

import Button from '../components/Button'
import ButtonBar from '../components/ButtonBar'
import Main from '../components/Main'
import Modal from '../components/Modal'
import { Text } from '../components/Typography'
import BallotContext from '../contexts/ballotContext'

const Header = styled.div`
  margin: 1rem 0;
`

const Table = styled.table`
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
`
interface TableCellProps {
  border?: boolean
}
const TableCell = styled.td`
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: ${({ border = false }: TableCellProps) =>
    border ? '1px solid lightGrey' : 'none'};
`

const SummaryPage = (props: RouteComponentProps) => {
  const { contests, resetBallot, votes } = useContext(BallotContext)
  const [isAlert, setAlert] = useState(false)
  const closeAlert = () => {
    setAlert(false)
  }
  const requestNewBallot = () => {
    Object.keys(votes).length === 0 ? startOver() : setAlert(true)
  }
  const startOver = () => {
    resetBallot()
    props.history.push('/')
  }
  return (
    <React.Fragment>
      <Main>
        <Header className="prose">
          <h1>Official Ballot</h1>
          <p className="no-print">
            Please review your ballot. Confirm your votes by selecting the
            “Print Ballot” button.
          </p>
        </Header>
        <Table>
          <caption className="no-print visually-hidden">
            <p>Summary of your votes.</p>
          </caption>
          <thead className="no-print">
            <tr>
              <TableCell as="th" scope="col">
                Contest
              </TableCell>
              <TableCell as="th" scope="col">
                Vote
              </TableCell>
            </tr>
          </thead>
          <tbody>
            {contests.map(contest => {
              const candidate = contest.candidates.find(
                c => c.name === votes[contest.id]
              )
              const vote = candidate ? (
                candidate.name
              ) : (
                <Text as="span" muted>
                  no selection
                </Text>
              )
              return (
                <tr key={contest.id}>
                  <TableCell as="th" border>
                    {contest.title}{' '}
                  </TableCell>
                  <TableCell border>
                    {vote}{' '}
                    <small className="no-print">
                      <Link to={`/contests/${contest.id}`}>change</Link>
                    </small>
                  </TableCell>
                </tr>
              )
            })}
          </tbody>
        </Table>
      </Main>
      <ButtonBar>
        <Button autoFocus onClick={window.print}>
          Print Ballot
        </Button>
        <Button onClick={props.history.goBack}>Back</Button>
        <Button onClick={requestNewBallot}>New Ballot</Button>
      </ButtonBar>
      <Modal isOpen={isAlert}>
        <Text>Clear all votes and start over?</Text>
        <button onClick={closeAlert}>Cancel</button>
        <button onClick={startOver}>Start Over</button>
      </Modal>
    </React.Fragment>
  )
}

export default SummaryPage
