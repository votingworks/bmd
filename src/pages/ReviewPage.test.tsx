import { axe } from 'jest-axe'
import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import SummaryPage from './ReviewPage'

it(`renders SummaryPage without votes`, () => {
  const { container } = render(
    <Route path="/review" component={SummaryPage} />,
    {
      route: '/review',
    }
  )
  expect(container.firstChild).toMatchSnapshot()
})

fit(`renders SummaryPage with votes`, () => {
  const { container } = render(
    <Route path="/review" component={SummaryPage} />,
    {
      route: '/review',
      votes: {
        president: 'Minnie Mouse',
        senator: 'John Smith',
      },
    }
  )
  expect(container.firstChild).toMatchSnapshot()
})

it(`SummaryPage without votes is accessible`, async () => {
  const resetBallot = jest.fn()
  const { container } = render(
    <Route path="/review" component={SummaryPage} />,
    {
      resetBallot,
      route: '/review',
    }
  )
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})

it(`SummaryPage with votes is accessible`, async () => {
  const resetBallot = jest.fn()
  const { container } = render(
    <Route path="/review" component={SummaryPage} />,
    {
      resetBallot,
      route: '/review',
      votes: {
        president: 'Minnie Mouse',
        senator: 'John Smith',
      },
    }
  )
  expect(await axe(container.innerHTML)).toHaveNoViolations()
})
