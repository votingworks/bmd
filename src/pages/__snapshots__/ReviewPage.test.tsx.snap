// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SummaryPage with votes 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: scroll;
  padding: 2rem 1rem;
}

.c1 {
  width: 100%;
  max-width: 40rem;
  margin: 0 auto;
}

.c2 {
  line-height: 1.3;
  max-width: 66ch;
}

.c2 h1 {
  line-height: 1.1;
  margin: 2rem 0 1rem;
}

.c2 h2 {
  margin: 1.5rem 0 1rem;
}

.c2 h3,
.c2 p {
  margin: 1rem 0;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p,
.c2 h3 + p {
  margin-top: -0.75rem;
}

.c2 > :first-child {
  margin-top: 0;
}

.c2 > :last-child {
  margin-bottom: 0;
}

.c5 {
  max-width: 320px;
  margin: auto;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  min-height: 9.5in;
  margin-top: 2rem;
  padding: 2rem;
  background: white;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 1rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 h2 {
  margin-bottom: 0;
}

.c4 h3 {
  margin-top: 0;
}

.c4 > div:nth-child(1) {
  width: 175px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

.c4 > div:nth-child(2) {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0 1rem;
}

.c10 {
  width: 50%;
  margin: 1rem 0 -0.75rem;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c7 {
  width: 100%;
  max-width: 66ch;
  text-align: left;
  border-bottom: 1px solid lightGrey;
}

.c8 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: none;
  font-weight: normal;
  line-height: 1.2;
  vertical-align: top;
}

.c9 {
  width: 50%;
  padding: 0.5rem 0.25rem;
  border-top: 1px solid lightGrey;
  font-weight: normal;
  line-height: 1.2;
  vertical-align: top;
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media print {
  .c1 {
    max-width: 100%;
    margin: 0;
  }
}

@media print {
  .c3 {
    margin: 0;
    padding: 0;
  }
}

@media print {
  .c7 {
    max-width: 100%;
  }
}

<main
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <h1
        class="no-print"
      >
        Review Your Selections
      </h1>
      <p
        class="no-print"
      >
        Confirm your votes by printing your ballot.
      </p>
    </div>
    <section
      class="c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <svg
            viewBox="0 0 500 500"
            xmlns="http://www.w3.org/2000/svg"
          >
            
  
            <circle
              cx="250"
              cy="250"
              fill="#fff"
              r="240"
              stroke="#000"
              stroke-width="10"
            />
            
  
            <text
              alignment-baseline="middle"
              style="font-family: sans-serif; font-size: 180%"
              text-anchor="middle"
              x="50%"
              y="50%"
            >
              Official Seal Here
            </text>
            

          </svg>
          

        </div>
        <div
          class="c2"
        >
          <h2>
            Official Ballot
          </h2>
          <h3>
            2020 General Election
          </h3>
          <p>
            Nearby County
            , 
            State of Somewhere
            <br />
            Tuesday, November 6, 2020
          </p>
        </div>
      </div>
      <div
        class="c6"
      >
        <table
          class="c7"
        >
          <caption
            class="no-print visually-hidden"
          >
            <p>
              Summary of your votes.
            </p>
          </caption>
          <thead>
            <tr>
              <th
                class="c8"
                scope="col"
              >
                Contest
              </th>
              <th
                class="c8"
                scope="col"
              >
                Vote
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <th
                class="c9"
              >
                President
                 
              </th>
              <td
                class="c9"
              >
                <strong>
                  Minnie Mouse
                </strong>
                 
                <small
                  class="no-print"
                >
                  <a
                    href="/contests/president"
                  >
                    change
                  </a>
                </small>
              </td>
            </tr>
            <tr>
              <th
                class="c9"
              >
                Senator
                 
              </th>
              <td
                class="c9"
              >
                <strong>
                  John Smith
                </strong>
                 
                <small
                  class="no-print"
                >
                  <a
                    href="/contests/senator"
                  >
                    change
                  </a>
                </small>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="c10"
      >
        <svg
          viewBox="0 0 1293 177"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h1293v177H0z"
            fill="#FFF"
          />
          <path
            d="M0 0h6v150H0zM9 0h3v150H9zM18 0h3v150h-3zM33 0h6v150h-6zM45 0h3v150h-3zM54 0h9v150h-9zM66 0h3v150h-3zM75 0h9v150h-9zM87 0h6v150h-6zM99 0h3v150h-3zM108 0h3v150h-3zM123 0h6v150h-6zM132 0h3v150h-3zM141 0h3v150h-3zM156 0h6v150h-6zM165 0h9v150h-9zM177 0h6v150h-6zM186 0h9v150h-9zM198 0h3v150h-3zM213 0h3v150h-3zM222 0h6v150h-6zM231 0h9v150h-9zM246 0h3v150h-3zM252 0h6v150h-6zM264 0h3v150h-3zM279 0h3v150h-3zM285 0h6v150h-6zM297 0h3v150h-3zM306 0h6v150h-6zM315 0h9v150h-9zM330 0h3v150h-3zM336 0h9v150h-9zM348 0h12v150h-12zM363 0h6v150h-6zM372 0h12v150h-12zM387 0h6v150h-6zM396 0h6v150h-6zM405 0h3v150h-3zM417 0h9v150h-9zM429 0h3v150h-3zM435 0h12v150h-12zM450 0h9v150h-9zM462 0h3v150h-3zM471 0h6v150h-6zM480 0h9v150h-9zM495 0h3v150h-3zM501 0h9v150h-9zM513 0h12v150h-12zM528 0h9v150h-9zM540 0h9v150h-9zM552 0h6v150h-6zM561 0h3v150h-3zM576 0h6v150h-6zM585 0h3v150h-3zM594 0h3v150h-3zM600 0h12v150h-12zM615 0h9v150h-9zM627 0h3v150h-3zM636 0h6v150h-6zM645 0h9v150h-9zM660 0h3v150h-3zM669 0h3v150h-3zM675 0h6v150h-6zM693 0h3v150h-3zM699 0h6v150h-6zM717 0h3v150h-3zM726 0h6v150h-6zM735 0h9v150h-9zM750 0h3v150h-3zM759 0h3v150h-3zM765 0h6v150h-6zM777 0h3v150h-3zM792 0h3v150h-3zM801 0h6v150h-6zM810 0h9v150h-9zM825 0h6v150h-6zM834 0h9v150h-9zM849 0h3v150h-3zM858 0h3v150h-3zM864 0h6v150h-6zM876 0h3v150h-3zM891 0h6v150h-6zM903 0h9v150h-9zM915 0h3v150h-3zM924 0h6v150h-6zM936 0h9v150h-9zM948 0h3v150h-3zM957 0h3v150h-3zM963 0h6v150h-6zM981 0h3v150h-3zM990 0h3v150h-3zM999 0h9v150h-9zM1011 0h6v150h-6zM1023 0h3v150h-3zM1032 0h3v150h-3zM1047 0h6v150h-6zM1056 0h3v150h-3zM1065 0h3v150h-3zM1080 0h6v150h-6zM1089 0h3v150h-3zM1104 0h3v150h-3zM1110 0h6v150h-6zM1122 0h9v150h-9zM1134 0h6v150h-6zM1143 0h9v150h-9zM1155 0h6v150h-6zM1167 0h3v150h-3zM1173 0h9v150h-9zM1188 0h3v150h-3zM1194 0h6v150h-6zM1206 0h3v150h-3zM1221 0h3v150h-3zM1227 0h6v150h-6zM1245 0h3v150h-3zM1254 0h6v150h-6zM1269 0h9v150h-9zM1281 0h3v150h-3zM1287 0h6v150h-6z"
          />
          <text
            font-family="Helvetica"
            font-size="24"
            text-anchor="middle"
            x="646.5"
            y="177"
          >
            40bb7d9c-9049-4873-af5e-5e66f0bbc73e
          </text>
        </svg>
      </div>
    </section>
  </div>
</main>
`;
